numpy/core/src/multiarray/getset.c

```diff
@@ -492,58 +490,70 @@ array_descr_set(PyArrayObject *self, PyObject *arg)
-         * dimension at end (or at the front for NPY_ARRAY_F_CONTIGUOUS)
-         */
-        if (PyArray_DESCR(self)->elsize % newtype->elsize != 0) {
+        else if (PyDataType_HASSUBARRAY(newtype)) {
+            PyErr_SetString(PyExc_ValueError,
+                    "Changing the dtype to a subarray type is only supported "
+                    "if the total itemsize is unchanged");
             goto fail;
         }
-        newdim = PyArray_DESCR(self)->elsize / newtype->elsize;
-        PyArray_DIMS(self)[i] *= newdim;
-        PyArray_STRIDES(self)[i] = newtype->elsize;
-    }
-    else if (newtype->elsize > PyArray_DESCR(self)->elsize) {
-        /*
-         * Determine if last (or first if NPY_ARRAY_F_CONTIGUOUS) dimension
-         * is compatible
-         */
-        newdim = PyArray_DIMS(self)[i] * PyArray_DESCR(self)->elsize;
-        if ((newdim % newtype->elsize) != 0) {
+
+        /* determine which axis to resize */
+        if (PyArray_IS_C_CONTIGUOUS(self)) {
+            axis = PyArray_NDIM(self) - 1;
+        }
+        else if (PyArray_IS_F_CONTIGUOUS(self)) {
+            /* 2015-11-27 1.11.0 */
+            if (DEPRECATE(
+                        "Changing the shape of an F-contiguous array by "
+                        "descriptor assignment is deprecated. To maintain the "
+                        "Fortran contiguity of a multidimensional Fortran "
+                        "array, use 'a.T.view(...).T' instead") < 0) {
+                goto fail;
+            }
+            axis = 0;
+        }
+        else {
+            /* Don't mention the deprecated F-contiguous support */
+            PyErr_SetString(PyExc_ValueError, "Array must be C-contiguous");
             goto fail;
         }
-        PyArray_DIMS(self)[i] = newdim / newtype->elsize;
-        PyArray_STRIDES(self)[i] = newtype->elsize;
+
+        if (newtype->elsize < PyArray_DESCR(self)->elsize) {
+            /* if it is compatible, increase the size of the relevant axis */
+            if (PyArray_DESCR(self)->elsize % newtype->elsize != 0) {
+                PyErr_SetString(PyExc_ValueError,
+                        "When changing to a smaller dtype, its size must be a "
+                        "divisor of the size of original dtype");
+                goto fail;
+            }
+            newdim = PyArray_DESCR(self)->elsize / newtype->elsize;
+            PyArray_DIMS(self)[axis] *= newdim;
+            PyArray_STRIDES(self)[axis] = newtype->elsize;
+        }
+        else if (newtype->elsize > PyArray_DESCR(self)->elsize) {
+            /* if it is compatible, decrease the size of the relevant axis */
+            newdim = PyArray_DIMS(self)[axis] * PyArray_DESCR(self)->elsize;
+            if ((newdim % newtype->elsize) != 0) {
+                PyErr_SetString(PyExc_ValueError,
+                        "When changing to a larger dtype, its size must be a "
+                        "multiple of the size of original dtype");
```

Comment:
Good catch, I'll update the message.

I don't think we can make it less loose without a deprecation cycle, and might be in favor of making the other case more loose (as in #6097). Having said that, I can't think of a single reason this would be useful)